### PR TITLE
Added new custom eslint rule to fix wallaby comments

### LINKED ISSUE #29: Add no-wallaby-comment rule
Error on wallaby-specific temporary comments like `// ?.` and `// file.only`

### PR DESCRIPTION
**Pull request by jpolavar**
Closes #29. Removes  wallaby-specific temporary comments like // ? or // ?? or // ?. or // file.only or // file.skip

### CODE REVIEW

File: .DS_Store
Comment (carlansley) [inline .DS_Store:1]: this file shouldn't be in the PR

File: src/no-wallaby-comment.spec.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    valid: [
+      {
+        code: LINE_WITH_NO_COMMENTS,
+      },
+    ],
+    invalid: [
+      {
+        code: LINE_WITH_COMMENT,
+        errors: [{ message: 'Remove wallaby-specific comments' }],
+        output: `const NOT_A_SECRET = "A template that isn't a secret."; `,
```
Comment (carlansley) [inline src/no-wallaby-comment.spec.ts:31]: i think the fix should remove all extra space from the end of the line

File: src/no-wallaby-comment.spec.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+      },
+      {
+        code: LINE_WITH_MULTIPLE_COMMENTS,
+        errors: [
+          { message: 'Remove wallaby-specific comments' },
+          { message: 'Remove wallaby-specific comments' },
+          { message: 'Remove wallaby-specific comments' },
+          { message: 'Remove wallaby-specific comments' },
+        ],
+        output: ` const NOT_A_SECRET = "A template that isn't a secret.";   `,
```
Comment (carlansley) [inline src/no-wallaby-comment.spec.ts:41]: same here

File: src/no-wallaby-comment.spec.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ *
+ * This code is licensed under the MIT license (see LICENSE.txt for details).
+ */
+import { RuleTester } from 'eslint';
+
+import rule from './no-wallaby-comment';
+
+const LINE_WITH_NO_COMMENTS = `const NOT_A_SECRET = "A template that isn't a secret.";`;
+
+const LINE_WITH_COMMENT = `const NOT_A_SECRET = "A template that isn't a secret."; // ?`;
```
Comment (carlansley) [inline src/no-wallaby-comment.spec.ts:14]: there's no test for `file.only`, and also i think there should be single comment tests for the various varieties (e.g. `// ?.` etc)

**Review by carlansley — CHANGES_REQUESTED** [review]
A few changes.  Another question, has this been tested on any of our code bases?  I don't see a beta of eslint-config that has this rule included.

**Comment by carlansley** [comment]
`file.skip` is another comment we're not dealing with here, there may be others also

File: src/no-wallaby-comment.spec.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+ * This code is licensed under the MIT license (see LICENSE.txt for details).
+ */
+import { RuleTester } from 'eslint';
+
+import rule from './no-wallaby-comment';
+
+const LINE_WITH_NO_COMMENTS = `const NOT_A_SECRET = "A template that isn't a secret.";`;
+
+const LINE_ONE_WITH_MULTIPLE_COMMENTS = `
+// file.skip const NOT_A_SECRET = "A template that isn't a secret.";
```
Comment (carlansley) [inline src/no-wallaby-comment.spec.ts:15]: file.skip and file.only are normally on lines by themselves, not in front of commented code.  think we need to test that scenario instead (tbh not sure what Wallabys behavior is here)

**Review by carlansley — CHANGES_REQUESTED** [review]

File: src/no-wallaby-comment.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    },
+    fixable: 'code',
+  },
+  create(context: Rule.RuleContext) {
+    const sourceCode = context.sourceCode;
+    const sourceLines = sourceCode.lines;
+
+    sourceLines.forEach((line, lineNumber) => {
+      const regex = /\s*\/\/\s*(?:\?{1,2}\.?|file\.(?:only|skip))\s*/gu;
+      const parts = line.split(regex);
```
[the remark this answers is not in the document]
Comment (carlansley) [reply src/no-wallaby-comment.ts]: has this been fixed/dealt with?  its unclear to me

**Comment by carlansley** [comment]
have you tried this on one of our larger internal repos?  Think it's worth trying, especially to make sure the regex is performant.

File: src/no-wallaby-comment.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jpolavar) [reply src/no-wallaby-comment.ts]: It was previous commit. It was fixed and we don't that specific functionality of splitting and regex in the latest commit

**Review by carlansley — CHANGES_REQUESTED** [review]
i played around with wallaby's comment system, noticed the following all work:
```
//       file.only

/* file.only    */

/*
 * file.only
 */
 
 /*
 * some other comment
 * file.skip
 */
```

i'm thinking we should make sure all of these various situations are identified/tested

**Comment by jpolavar** [comment]
> i played around with wallaby's comment system, noticed the following all work:
> 
> ```
> //       file.only
> 
> /* file.only    */
> 
> /*
>  * file.only
>  */
>  
>  /*
>  * some other comment
>  * file.skip
>  */
> ```
> 
> i'm thinking we should make sure all of these various situations are identified/tested
Tested different scenarios on aws-nock and test cases as well to the library

**Review by carlansley — CHANGES_REQUESTED** [review]

**Review by carlansley — APPROVED** [review]

**Pull request merged**